src/main/java/org/sagebionetworks/bridge/services/PermissionService.java

```diff
@@ -158,6 +172,120 @@ public void deletePermission(String appId, String guid) {
+                        entityTypeMap.put(entityId, additionalAccessLevels);
+                        additionalPermissions.put(entityType, entityTypeMap);
+                    } else {
+                        Map<String, Set<AccessLevel>> entityTypeMap = additionalPermissions.getOrDefault(entityType, new HashMap<>());
+                        entityTypeMap.put(entityId, newPermissionEntityTypeMap.get(entityId));
+                        additionalPermissions.put(entityType, entityTypeMap);
+                    }
+                }
+            } else {
+                // the old permissions do not have any of the new so add them all
+                additionalPermissions.put(entityType, newPermissions.get(entityType));
+            }
+        }
+        
+        for (EntityType entityType : oldPermissions.keySet()) {
+            if (newPermissions.containsKey(entityType)) {
+                Map<String, Set<AccessLevel>> oldPermissionEntityTypeMap = newPermissions.get(entityType);
+                for (String entityId : oldPermissionEntityTypeMap.keySet()) {
+                    if (newPermissions.get(entityType).containsKey(entityId)) {
+                        Set<AccessLevel> newAccessLevels = newPermissions.get(entityType).get(entityId);
+                        Set<AccessLevel> removableAccessLevels = new HashSet<>();
+                        for (AccessLevel accessLevel : oldPermissionEntityTypeMap.get(entityId)) {
+                            if (!newAccessLevels.contains(accessLevel)) {
+                                removableAccessLevels.add(accessLevel);
+                            }
+                        }
+                        Map<String, Set<AccessLevel>> entityTypeMap = removablePermissions.getOrDefault(entityType, new HashMap<>());
+                        entityTypeMap.put(entityId, removableAccessLevels);
+                        removablePermissions.put(entityType, entityTypeMap);
+                    } else {
+                        Map<String, Set<AccessLevel>> entityTypeMap = removablePermissions.getOrDefault(entityType, new HashMap<>());
+                        entityTypeMap.put(entityId, oldPermissionEntityTypeMap.get(entityId));
+                        removablePermissions.put(entityType, entityTypeMap);
+                    }
+                }
+            } else {
+                removablePermissions.put(entityType, oldPermissions.get(entityType));
+            }
+        }
+        
+        List<Permission> currentPermissions = permissionDao.getPermissionsForUser(account.getAppId(), account.getId());
+        for (EntityType entityType : removablePermissions.keySet()) {
+            for (String entityId : removablePermissions.get(entityType).keySet()) {
+                for (AccessLevel accessLevel : removablePermissions.get(entityType).get(entityId)) {
+                    for (Permission permission : currentPermissions) {
```

Comment:
Looping through all existing permissions for each entity type, id, and access level is very inefficient. It might be simplier to loop through permissions as the outermost loop, and then getting the entity type, id, and access level from the permission and calling the getters in the map. This way, we don't need nested loops to do this operation.

Reply:
That makes sense. Updated to loop through existing permissions and delete where there is a match.